src/node_jsvmapi.cc

```diff
@@ -1569,122 +1600,153 @@ napi_status napi_coerce_to_string(napi_env e, napi_value v, napi_value* result)
 }
 
-napi_status napi_release_persistent(napi_env e, napi_persistent p) {
+napi_status napi_get_value_external(napi_env e, napi_value v, void** result) {
   NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
 
-  v8::Persistent<v8::Value> *thePersistent =
-      v8impl::V8PersistentValueFromJsPersistentValue(p);
-  thePersistent->Reset();
-  delete thePersistent;
+  v8::Local<v8::External> externalValue = v8impl::V8LocalValueFromJsValue(v).As<v8::External>();
+  *result = externalValue->Value();
 
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_get_persistent_value(napi_env e, napi_persistent p, napi_value* result) {
+
+// Set initial_refcount to 0 for a weak reference, >0 for a strong reference.
+// The finalizer callback and callback data parameters optional.
+napi_status napi_create_reference(napi_env e,
+                                  napi_value value,
+                                  int initial_refcount,
+                                  napi_ref* result) {
   NAPI_PREAMBLE(e);
   CHECK_ARG(result);
+  RETURN_STATUS_IF_FALSE(initial_refcount >= 0, napi_invalid_arg);
 
   v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Persistent<v8::Value> *thePersistent =
-      v8impl::V8PersistentValueFromJsPersistentValue(p);
-  v8::Local<v8::Value> napi_value =
-      v8::Local<v8::Value>::New(isolate, *thePersistent);
 
-  *result = v8impl::JsValueFromV8LocalValue(napi_value);
+  v8impl::Reference* reference = new v8impl::Reference(
+    isolate, v8impl::V8LocalValueFromJsValue(value), initial_refcount, false);
+
+  *result = reinterpret_cast<napi_ref>(reference);
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_create_weakref(napi_env e, napi_value v, napi_weakref* result) {
+// Deletes a reference. The referenced value is released, and may be GC'd unless there
+// are other references to it.
+napi_status napi_delete_reference(napi_env e, napi_ref ref) {
   NAPI_PREAMBLE(e);
-  CHECK_ARG(result);
+  CHECK_ARG(ref);
+
+  v8impl::Reference* reference = reinterpret_cast<v8impl::Reference*>(ref);
+  delete reference;
 
-  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Persistent<v8::Value> *thePersistent =
-      new v8::Persistent<v8::Value>(
-          isolate, v8impl::V8LocalValueFromJsValue(v));
-  thePersistent->SetWeak(static_cast<int*>(nullptr), v8impl::WeakRefCallback,
-                         v8::WeakCallbackType::kParameter);
-  // need to mark independent?
-  *result = v8impl::JsWeakRefFromV8PersistentValue(thePersistent);
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_get_weakref_value(napi_env e, napi_weakref w, napi_value* result) {
+// Increments the reference count, optionally returning the resulting count. After this call the
+// reference will be a strong reference because its refcount is >0, and the referenced object is
+// effectively "pinned". Calling this when the refcount is 0 and the object is unavailable
```

Comment:
Is this true about an error when the refcount is 0 and the object is unavailable? Is this an exception thrown by ClearWeak?